xarray/backends/zarr.py

```diff
@@ -0,0 +1,554 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+import functools
+from itertools import product
+# import operator
+
+import numpy as np
+
+from .. import Variable
+from ..core import indexing
+from ..core.utils import FrozenOrderedDict, HiddenKeyDict
+from ..core.pycompat import iteritems, OrderedDict
+
+from .common import (AbstractWritableDataStore, DataStorePickleMixin,
+                     BackendArray)
+
+from .. import conventions
+
+# this is a private method but we need it for open_zar
+from .api import _protect_dataset_variables_inplace
+
+# need some special secret attributes to tell us the dimensions
+_DIMENSION_KEY = '_ARRAY_DIMENSIONS'
+
+
+# zarr attributes have to be serializable as json
+# many xarray datasets / variables have numpy arrays and values
+# these functions handle encoding / decoding of such items
+def _encode_zarr_attr_value(value):
+    # what is the most duck-type friendly way to do this check
+    if isinstance(value, np.ndarray):
+        return value.tolist()
+    # I don't know how to check generically if something is a numpy scalar
+    # i.e. np.float32 or np.int8, etc. without checking against each dtype
+    # manually. This was the best I could come up with
+    elif isinstance(value, np.generic):
+        # np.string_('X').item() returns a type `bytes`
+        # zarr still doesn't like that
+        # causes some fill_value encoding to fail
+        return value.item()
+    else:
+        return value
+
+
+def _ensure_valid_fill_value(value, dtype):
+    if dtype.type == np.string_ and type(value) == bytes:
+        valid = value.decode('ascii')
+    else:
+        valid = value
+    return _encode_zarr_attr_value(valid)
+
+
+# TODO: cleanup/combine these next two functions
+def _decode_zarr_attr_value(value):
+    # what happens if we just don't decode anything?
+    # does it matter that we don't convert back to numpy types?
+    return value
+
+
+def _decode_zarr_attrs(attrs):
+    return OrderedDict([(k, _decode_zarr_attr_value(v))
+                        for k, v in attrs.items()])
+
+
+# untested, but I think this does the appropriate shape munging to make slices
```

Comment:
We really should write some unit tests for this... I can probably help.

Reply:
I did not address this comment yet because it don't fully understand what the function does. @jhamman wrote that part for compatibility with the new orthogonal indexing stuff.